lib/apiservers/service/restapi/handlers/vch_delete.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handlers
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/go-openapi/runtime/middleware"
+
+	"github.com/vmware/vic/lib/apiservers/service/models"
+	"github.com/vmware/vic/lib/apiservers/service/restapi/handlers/util"
+	"github.com/vmware/vic/lib/apiservers/service/restapi/operations"
+	"github.com/vmware/vic/lib/install/data"
+	"github.com/vmware/vic/lib/install/management"
+	"github.com/vmware/vic/lib/install/validate"
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/pkg/version"
+)
+
+// VCHDelete is the handler for deleting a VCH
+type VCHDelete struct {
+}
+
+// VCHDatacenterDelete is the handler for deleting a VCH within a Datacenter
+type VCHDatacenterDelete struct {
+}
+
+func (h *VCHDelete) Handle(params operations.DeleteTargetTargetVchVchIDParams, principal interface{}) middleware.Responder {
+	op := trace.NewOperation(params.HTTPRequest.Context(), "VCHDelete: %s", params.VchID)
+
+	b := buildDataParams{
+		target:     params.Target,
+		thumbprint: params.Thumbprint,
+	}
+
+	d, err := buildData(op, b, principal)
+	if err != nil {
+		return operations.NewDeleteTargetTargetVchVchIDDefault(util.StatusCode(err)).WithPayload(&models.Error{Message: err.Error()})
+	}
+
+	d.ID = params.VchID
+
+	err = deleteVCH(op, d, params.DeletionSpecification)
```

Comment:
This can be collapsed to:

```go
if err = deleteVCH(op, d, params.DeletionSpecification); err != nil {
    ...
}
```

Reply:
It can be collapsed, but I don't think doing so improves readability.

With the current structure, `err` is not as tightly scoped as it could be, but the overall flow is clear. By collapsing the assignments into initialization statements of the conditional, I think it becomes harder to read the overall flow of the function.

I think it's best to limit our use of optional initialization statements to calls where the sole purpose of the statement is the conditional. Typically, these are calls which are free of side-effects.

This allows someone skimming the code to easily see what's happening. In the following example, it's clear that the overall flow is to check something, do something (and handle any errors), validate something, call a method, and then do something else (and handle any errors). 
```
if total := a+b+c; total <= 0 {
   ...
}

d, err := doThingOne(a, b, c)
if err != nil {
   ...
}

if err = validateResult(d); err != nil {
   ...
}

d.callThisMethod()

err = doThingThree(e)
if err != nil {
    ...
}
```

We can re-write this, collapsing more aggressively, and have the same behavior.
```
if total := a+b+c; total <= 0 {
   ...
}

d, err := doThingOne(a, b, c)
if err != nil {
   ...
}

if err = validateResult(d); err != nil {
   ...
}

d.callThisMethod()

if err = doThingThree(e); err != nil {
    ...
}
```
However, this structure emphasizes the conditional nature of `doThingThree` (i.e., that it may result in an error) instead of emphasizing that we're doing something (and then handling an error).

The difference is subtle, but I think it improves readability.

Are there advantages to the collapsed version I may be overlooking?

Reply:
Sounds good! I brought this up purely for idiomatic purposes - if variables assigned in an `if` block aren't used outside of it and an error is being checked in the said block, we can perform the assignment and the check on the same line. However, if it makes more sense to separate them for readability, we can do that.